### ISSUE TITLE
Add Texture and Sampling Hardware Support

### ISSUE THREAD

**Issue by ifsheldon** [taichi-dev/taichi#2380]
**Concisely describe the proposed feature**
Please consider design APIs for accessing texture memories and utilizing hardware-accelerated samplers.

**Describe the solution you'd like (if any)**
As far as I know, backends like CUDA and OpenGL have dedicated texture memory handling functions and hardware acceleration for texture fetch and sampling. But now, samplings in `taichi_glsl` is only soft implementations, I think. So, it will be great if you can add APIs that directly interact with the underlying hardware accelerated backend APIs.

**Comment by maajor** [comment]
+1

Please see [this perftest](https://github.com/sebbbi/perftest) for different type of buffer comparison.

Seems most buffer differ not much, but bilinear interpolation on Texture2D do make a difference. See the benchmark that nearest sample, bilinear sample and load on Texture2D take similar time.

As for mutability, DirectX is the same, you can declare with RW or read-only for example, but I think it need more profiling to tell if there is any difference.

**Comment by k-ye** [comment]
FYI, we are tracking this work in https://github.com/taichi-dev/taichi/issues/5109

**Comment by ifsheldon** [comment]
Thanks for the update! 

How about mutability hints? Do you have any plans of adding them? Mutability hints may be beneficial for both performance and debugging.

**Comment by k-ye** [comment]
I think mutability hints and texture support are two different things, and #5109 is talking about the later :-) But yeah, we could leverage constant memory to pass read-only data. From the mutability perspective, we only intend to support texture sampling (read-only operations) as a start, if that's what you are asking...

**Comment by ifsheldon** [comment]
> From the mutability perspective, we only intend to support texture sampling (read-only operations) as a start, if that's what you are asking...

Yeah, but not only texture sampling. I mean, generally, if data is asserted to be read-only, some helpful performance optimization and checks can be (more easily) implemented. For example, I remember there're `__ldg` intrinsics and simple constant pointer hints for CUDA users to hint the NVCC that some data are read-only and accessed with only read operations. Probably you can let Taichi users to hint or assert that a field that is passed to a kernel is read-only so then any write operations to the read-only field is forbidden. This can help users to prevent some read-write bugs and help you Taichi developers to optimize the performance (say, loading read-only data from a field using `__ldg` when CUDA backend is used, or use constant memory if applicable).
